reactor-netty-incubator-quic/src/main/java/reactor/netty/incubator/quic/QuicClient.java

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright (c) 2021 VMware, Inc. or its affiliates, All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package reactor.netty.incubator.quic;
+
+import reactor.core.publisher.Mono;
+import reactor.netty.transport.AddressUtils;
+
+import java.net.SocketAddress;
+import java.time.Duration;
+import java.util.Objects;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+/**
+ * A QuicClient allows building in a safe immutable way a QUIC client that is materialized
+ * and connecting when {@link #connect()} is ultimately called.
+ * <p>
+ * <p> Example:
+ * <pre>
+ * {@code
+ *     QuicClient.create()
+ *               .port(7777)
+ *               .bindAddress(() -> new InetSocketAddress(0))
+ *               .wiretap(true)
+ *               .secure(clientCtx)
+ *               .idleTimeout(Duration.ofSeconds(5))
+ *               .initialSettings(spec ->
+ *                   spec.maxData(10000000)
+ *                       .maxStreamDataBidirectionalLocal(1000000))
+ *               .connectNow();
+ * }
+ * </pre>
+ *
+ * @author Violeta Georgieva
+ */
+public abstract class QuicClient extends QuicTransport<QuicClient, QuicClientConfig> {
+
+	/**
+	 * Prepare a {@link QuicClient}
+	 *
+	 * @return a {@link QuicClient}
+	 */
+	public static QuicClient create() {
+		return QuicClientConnect.INSTANCE;
+	}
+
+	/**
+	 * Connect the {@link QuicClient} and return a {@link Mono} of {@link QuicConnection}. If
+	 * {@link Mono} is cancelled, the underlying connection will be aborted. Once the
+	 * {@link QuicConnection} has been emitted and is not necessary anymore, disposing must be
+	 * done by the user via {@link QuicConnection#dispose()}.
+	 *
+	 * @return a {@link Mono} of {@link QuicConnection}
+	 */
+	public abstract Mono<? extends QuicConnection> connect();
```

Comment:
is the `< extends` necessary? usually, we're precise in return types and only use wildcards for input parameters